theme.js

```diff
@@ -1,117 +1,84 @@
 /* jshint node: true */
 'use strict';
 
+var path = require('path');
+
 var Addon = require('ember-cli/lib/models/addon');
+var ThemeCore = require('./lib/theme-core');
+var TransformComponentClasses = require('./lib/htmlbars-plugins/transform-component-classes');
+var TransformUITableComponents = require('./lib/htmlbars-plugins/transform-ui-table-components');
+var funnel = require('broccoli-funnel');
 
-var path = require('path');
-var csso = require('broccoli-csso');
+var themeCore;
 
-var jsPreprocessor = require('./lib/js-preprocessor');
-var templatePreprocessor = require('./lib/template-preprocessor');
-var scssPreprocessor = require('./lib/scss-preprocessor');
+var Theme = Addon.extend({
+  themeCore: null,
+  parentTheme: null,
 
-var TransformComponentClasses = require('./lib/transform-component-classes');
-var TransformUITableComponents = require('./lib/plugins/transform-ui-table-components');
+  isDevelopingAddon: function() {
+    return true;
+  },
 
-var UITheme = Addon.extend({
   init: function() {
     this._super.init.apply(this, arguments);
-    this._originalCompileStyles = this.compileStyles;
-    this.compileStyles = function() {};
-  },
 
-  treeForStyles() {
-    return this._originalCompileStyles(this.treeGenerator(this.root + '/addon/styles'));
+    this.themeCore = themeCore || new ThemeCore();
+    themeCore = this.themeCore;
+
+    this.parentTheme = this.themeCore.parentTheme;
+    this.themeCore.parentTheme = this;
+
+    this.themeCore.register(this);
   },
 
   setupPreprocessorRegistry: function(type, registry) {
-    this.setupJsPreprocessing(registry, type);
+    // only setup preprocessors for the top level app registry
+    // and not for subthemes that are parents to baser themes
+    if (type === 'parent' && !registry.app.parent) {
+      this.themeCore.setupAppPreprocessors(registry);
+    }
 
     if (type === 'self') {
-      this.setupTemplatePreprocessing(registry);
-      this.setupCssPreprocessing(registry);
-      this.setupHtmlTransform(registry);
+      this.setupHtmlTransforms(registry);
```

Comment:
there might not be any good reason why we do this for each `self` rather than aggregating it like the others